PyTorch Lightning: Add newer versions

kwryankrattiger opened the pull request:
<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

kwryankrattiger committed:
PyTorch Lightning: Add newer versions

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

adamjstewart reviewed (commented):
Also need to update dependencies. Note that py-lightning is basically identical to py-pytorch-lightning. I switched to the former and haven't been updating the latter, although neither has been officially deprecated.

kwryankrattiger committed:
Update dependencies for pytorch-lighting

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
Drop 2.5.5 due to URL changes

Skipping fixing this for now, dependencies are updated for 2.5: so
adding it back should be a matter of fixing the URL

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger commented:
@adamjstewart I fixed the deps. But version >=2.5 have changed the URL for the source so that is going to require additional effort and I currently don't need that version so I am punting in this PR.

I did leave the version constraint updates, so whoever does eventually need that version or later can handle the URL and the rest should be done.

adamjstewart commented on repos/spack_repo/builtin/packages/py_pytorch_lightning/package.py:
```diff
@@ -57,8 +64,13 @@ class PyPytorchLightning(PythonPackage):
     depends_on("py-setuptools", type=("build", "run"))
```
build/run for lightning 2.3 and older, build-only for 2.4 and newer

adamjstewart commented on repos/spack_repo/builtin/packages/py_pytorch_lightning/package.py:
```diff
@@ -57,8 +64,13 @@ class PyPytorchLightning(PythonPackage):
+    depends_on("py-numpy@1.17.2:", when="@1.3:2.3", type=("build", "run"))
     depends_on("py-numpy@1.16.6:", when="@:1.2", type=("build", "run"))
+    depends_on("py-torch@2.1:", when="@2.4:", type=("build", "run"))
+    depends_on("py-torch@2.1:", when="@2.4:", type=("build", "run"))
```
Duplicated

adamjstewart commented on repos/spack_repo/builtin/packages/py_pytorch_lightning/package.py:
```diff
@@ -67,23 +79,30 @@ class PyPytorchLightning(PythonPackage):
     depends_on("py-tqdm@4.57.0:", when="@1.6.3:", type=("build", "run"))
     depends_on("py-tqdm@4.41.0:", when="@:1.6.2", type=("build", "run"))
-    depends_on("py-pyyaml@5.4:", when="@1.6:", type=("build", "run"))
+    depends_on("py-pyyaml@5.4.1:", when="@1.6:", type=("build", "run"))
```
```suggestion
    depends_on("py-pyyaml@5.4.1:", when="@2.5.3:", type=("build", "run"))
    depends_on("py-pyyaml@5.4:", when="@1.6:", type=("build", "run"))
```

adamjstewart reviewed (commented):
Only checked 2.4 since that's the default. This package apparently doesn't have upper bounds on deps like lightning has. Otherwise, we could share the dep info like we do with py-pillow and py-pillow-simd.
